src/portable/raspberrypi/rp2040/rp2040_usb.c

```diff
@@ -57,6 +55,36 @@ TU_ATTR_ALWAYS_INLINE static inline bool is_host_mode(void)
 //
 //--------------------------------------------------------------------+
 
+
+uint32_t rp2040_critical_frame_period(struct hw_endpoint *ep)
+{
+#if TUD_OPT_RP2040_USB_DEVICE_UFRAME_FIX
+  uint32_t now;
+
+  if (usb_hw->main_ctrl & USB_MAIN_CTRL_HOST_NDEVICE_BITS)
+    return 0;
+
+  if (tu_edpt_dir(ep->ep_addr) == TUSB_DIR_OUT ||
+      ep->transfer_type == TUSB_XFER_INTERRUPT ||
+      ep->transfer_type == TUSB_XFER_ISOCHRONOUS)
+    return 0;
+
+  /* Avoid the last 200us (uframe 6.5-7) of a frame, up to the EOF2 point.
+   * The device state machine cannot recover from receiving an incorrect PID
+   * when it is expecting an ACK.
+   */
+  now = time_us_32();
+  if (now < (last_sof + 800) || now > (last_sof + 998)) {
```

Comment:
i would do
```
delta =  time_us_32() - last_sof
```
to avoid rollover issues (not a biggy but still)

Reply:
Rollover would mean the time check would not apply for one frame in every 4.29 million, which is arguably a bug.